src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -725,16 +725,17 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
 
         // Async mode - add vector to the frontend index and schedule an update job if needed.
         if (!this->backendIndex->isMultiValue()) {
+            {
+                std::shared_lock lock(this->flatIndexGuard);
+                // If the label already exists in the frontend index, we should count it
+                // to prevent the case when existing vector is moved meanwhile by the update job.
+                if (this->frontendIndex->isLabelExists(label)) {
+                    ret = -1;
+                }
```

Comment:
Can we move this logic to after line 741, so we will not have to acquire the flat index lock twice?

Reply:
There is still possibile that the vector can be moved between lines 735-737, so we can count it twice, and `ret==-2`.
But this case is covered by `std::max()` call at line 751

Reply:
I understand. So there is no way to avoid this additional locking?